crates/flipperzero/src/furi/string.rs

```diff
@@ -0,0 +1,259 @@
+//! String primitives built around `FuriString`.
+
+use core::{
+    cmp::Ordering,
+    convert::Infallible,
+    ffi::{c_char, CStr},
+    fmt, ops,
+};
+
+#[cfg(feature = "alloc")]
+use alloc::ffi::CString;
+
+use flipperzero_sys as sys;
+
+/// A Furi string.
+///
+/// This is similar to Rust's [`String`], but used inside the Flipper Zero SDK.
+#[derive(Eq)]
+pub struct String(*mut sys::FuriString);
+
+impl Drop for String {
+    fn drop(&mut self) {
+        unsafe { sys::furi_string_free(self.0) };
+    }
+}
+
+// Implementations matching `std::string::String`.
+impl String {
+    /// Creates a new empty `String`.
+    pub fn new() -> Self {
+        String(unsafe { sys::furi_string_alloc() })
+    }
+
+    /// Creates a new empty `String` with at least the specified capacity.
+    pub fn with_capacity(capacity: usize) -> Self {
+        let mut s = Self::new();
+        s.reserve(capacity);
+        s
+    }
+
+    /// Extracts a string slice containing the entire `String`.
+    #[inline]
+    #[must_use]
+    pub fn as_str(&self) -> &Str {
+        Str::from_raw(self.0)
+    }
+
+    /// Appends a given `CStr` onto the end of this `String`.
+    pub fn push_cstr(&mut self, string: &CStr) {
+        unsafe { sys::furi_string_cat_str(self.0, string.as_ptr()) }
+    }
+
+    /// Appends a given string slice onto the end of this `String`.
+    pub fn push_str(&mut self, string: &str) {
+        self.reserve(string.len());
+        for ch in string.chars() {
+            self.push(ch);
+        }
+    }
+
+    /// Reserves capacity for at least `additional` bytes more than the current length.
+    pub fn reserve(&mut self, additional: usize) {
+        unsafe { sys::furi_string_reserve(self.0, self.len() + additional) };
```

Comment:
Thanks. The documentation of it was unclear, and I read it as meaning that the `nul` character was handled for you (since every other method appears to handle the `nul`).

Reply:
Wait, hang on. Doesn't that mean `sys::furi_string_size` returns the length including the `nul` character? In which case this method is fine, and the bug is that `String::len` requires an additional `- 1`?

Reply:
Okay, [this part of the contract check](https://sourcegraph.com/github.com/P-p-H-d/mlib@62c8ac3e5d4a7a4f8757328e7a80286fde2686b6/-/blob/m-string.h?L40-41) appears to indicate that `string_size` returns the length of the string up to, but not including, the `nul` character.

And [this part of `string_reserve`](https://sourcegraph.com/github.com/P-p-H-d/mlib@62c8ac3e5d4a7a4f8757328e7a80286fde2686b6/-/blob/m-string.h?L318-322) ensures that the requested `alloc` is at least `size + 1`, which confirms that `alloc` needs to count the `nul` character.

Reply:
I hate looking at off-by-one issues like this. I got such a headache last night. 😄